ttnn/cpp/ttnn/operations/ccl/reduce_scatter/device/reduce_scatter_op.cpp

```diff
@@ -9,6 +9,55 @@
 
 namespace ttnn {
 
+ReduceScatter create_reduce_scatter_struct (
+    const Tensor& input_tensor,
+    const ttnn::operations::binary::BinaryOpType binary_op_type,
+    const uint32_t scatter_dim,
+    const uint32_t num_links,
+    const MemoryConfig output_mem_config,
+    const std::optional<size_t> user_defined_num_workers,
+    const std::optional<size_t> user_defined_num_buffers_per_channel,
+    const std::vector<Device*>& devices,
+    const ttnn::ccl::Topology topology
+){
+    uint32_t num_devices = devices.size();
+
+    bool is_linear = topology == ttnn::ccl::Topology::Linear;
+
+    uint32_t device_index = 0; // Initialize device index
+    std::optional<chip_id_t> receiver_device_id = std::nullopt; // Initialize receiver device ID
+    std::optional<chip_id_t> sender_device_id = std::nullopt; // Initialize sender device ID
+    for (uint32_t i = 0; i < num_devices; ++i) {
+        if (devices.at(i) == input_tensor.device()) {
+
+            bool is_last_chip_in_clockwise_direction = is_linear && i == (num_devices - 1);
+            bool is_last_chip_in_counter_clockwise_direction = is_linear && i == 0;
+            device_index = i;
+            receiver_device_id = is_last_chip_in_clockwise_direction ?
+                std::nullopt :
+                std::optional<chip_id_t>(devices.at((i + 1) % num_devices)->id());
+            sender_device_id = is_last_chip_in_counter_clockwise_direction ?
+                std::nullopt :
+                std::optional<chip_id_t>(devices.at((i + num_devices - 1) % num_devices)->id());
+            break;
+        }
+    }
+    TT_FATAL(receiver_device_id != std::nullopt || sender_device_id != std::nullopt, "Error, Reduce-scatter was unable to identify either a sender or receiver device ID and atleast one must be identified for a valid Reduce-scatter configuration. The input mesh tensor or Reduce-scatter arguments may be incorrect");
```

Comment:
This block of code is confusing to me:
- If `is_linear` is False, this block of code does nothing and recv/send device id is are both `std::nullopt`?
- `TT_FATAL` is checking that both recv/send device id are `std::nullopt`?
- What happens if `num_devices=1`?

Maybe I am completely misunderstanding what this code does :D

Reply:
tests will be skipped if num_devices=1, so it wont enter these part